qiskit/scheduler/methods/basic.py

```diff
@@ -140,31 +142,50 @@ def translate_gates_to_pulse_defs(circuit: QuantumCircuit,
     circ_pulse_defs = []
 
     cmd_def = schedule_config.cmd_def
-    measured_qubits = set()  # Collect qubits that would like to be measured
+    qubit_mem_slots = {}  # Map measured qubit index to classical bit index
 
     def get_measure_schedule() -> CircuitPulseDef:
         """Create a schedule to measure the qubits queued for measuring."""
         measures = set()
         all_qubits = set()
         sched = Schedule()
-        for q in measured_qubits:
-            measures.add(tuple(schedule_config.meas_map[q]))
+        for qubit in qubit_mem_slots:
+            measures.add(tuple(schedule_config.meas_map[qubit]))
         for qubits in measures:
             all_qubits.update(qubits)
-            # TODO (Issue #2704): Respect MemorySlots from the input circuit
-            sched |= cmd_def.get('measure', qubits)
-        measured_qubits.clear()
+            unused_mem_slots = set(qubits) - set(qubit_mem_slots.values())
+            default_sched = cmd_def.get('measure', qubits)
+            for time, inst in default_sched.instructions:
+                if isinstance(inst, AcquireInstruction):
+                    mem_slots = []
+                    for channel in inst.acquires:
+                        if channel.index in qubit_mem_slots.keys():
+                            mem_slots.append(MemorySlot(qubit_mem_slots[channel.index]))
+                        else:
+                            mem_slots.append(MemorySlot(unused_mem_slots.pop()))
+                    new_acquire = AcquireInstruction(command=inst.command,
+                                                     acquires=inst.acquires,
+                                                     mem_slots=mem_slots)
+                    sched._union((time, new_acquire))
+                # Measurement pulses should only be added if its qubit was measured by the user
+                elif inst.channels[0].index in qubit_mem_slots.keys():
+                    sched._union((time, inst))
+        qubit_mem_slots.clear()
         return CircuitPulseDef(schedule=sched, qubits=list(all_qubits))
 
-    for inst, qubits, _ in circuit.data:
+    for inst, qubits, clbits in circuit.data:
         inst_qubits = [qubit.index for qubit in qubits]  # We want only the indices of the qubits
-        if any(q in measured_qubits for q in inst_qubits):
+        if any(q in qubit_mem_slots for q in inst_qubits):
             # If we are operating on a qubit that was scheduled to be measured, process that first
             circ_pulse_defs.append(get_measure_schedule())
         if isinstance(inst, Barrier):
             circ_pulse_defs.append(CircuitPulseDef(schedule=inst, qubits=inst_qubits))
         elif isinstance(inst, Measure):
-            measured_qubits.update(inst_qubits)
+            if (len(inst_qubits) != 1 and len(clbits) != 1):
```

Comment:
Won't this cause a failure if the number of qubits != classicalbit and > 1. Should this be `or`?

Reply:
yes, shoot, I didn't notice this section had changed. I'll update it